src/libraries/Microsoft.Extensions.Logging.Console/src/StringWriter.cs

```diff
@@ -0,0 +1,182 @@
+        {
+            _stringBuilder = new StringBuilder();
+            DisableColors = disableColors;
+        }
+
+        public override void Write(char value)
+        {
+            _stringBuilder.Append(value);
+        }
+
+        public bool SetBackgroundColor(ConsoleColor? background)
+        {
+            if (DisableColors || background == _colorSettings.bg)
+            {
+                return false;
+            }
+            _colorSettings.bg = background;
+            if (background.HasValue)
+            {
+                Write(GetBackgroundColorEscapeCode(background.Value));
+            }
+            else
+            {
+                Write(DefaultBackgroundColor);
+            }
+            return true;
+        }
+
+        public bool SetForegroundColor(ConsoleColor? foreground)
+        {
+            if (DisableColors || foreground == _colorSettings.fg)
+            {
+                return false;
+            }
+            _colorSettings.fg = foreground;
+            if (foreground.HasValue)
+            {
+                Write(GetForegroundColorEscapeCode(foreground.Value));
+            }
+            else
+            {
+                Write(DefaultForegroundColor);
+            }
+            return true;
+        }
+
+        private (ConsoleColor? bg, ConsoleColor? fg) _colorSettings = (null, null);
+
+        public void ResetColor()
+        {
+            SetBackgroundColor(null);
+            SetForegroundColor(null);
+        }
+
+        private const string DefaultForegroundColor = "\x1B[39m\x1B[22m"; // reset to default foreground color
+        private const string DefaultBackgroundColor = "\x1B[49m"; // reset to the background color
+
+        public void WriteReplacing(string oldValue, string newValue, string message)
+        {
+            int len = _stringBuilder.Length;
+            _stringBuilder.Append(message);
+            _stringBuilder.Replace(oldValue, newValue, len, message.Length);
+        }
+
+        public int Length
+        {
+            get
+            {
+                return _stringBuilder.Length;
+            }
+        }
+
+        private string _computed;
+        public string ComputeAnsiString()
+        {
+            if (_stringBuilder.Length != 0)
+            {
+                ResetColor();
+                _computed = _stringBuilder.ToString();
+                _stringBuilder.Clear();
+            }
+            return _computed;
+        }
+
+        private void Clear()
+        {
+            if (_stringBuilder.Length != 0)
+            {
+                ResetColor();
+                _computed = _stringBuilder.ToString();
+                _stringBuilder.Clear();
+            }
+            if (_stringBuilder.Capacity > 1024)
+            {
+                _stringBuilder.Capacity = 1024;
+            }
+        }
+
+        private static string GetForegroundColorEscapeCode(ConsoleColor color)
```

Comment:
>GetForegroundColorEscapeCode [](start = 30, length = 28)

how this is different than what we already have in AnsiLogConsole.cs? we can refactor the code and have one copy.